tests/StackExchange.Redis.Tests/GarbageCollectionTests.cs

```diff
@@ -52,4 +53,43 @@ public async Task MuxerIsCollected()
 //            Assert.Equal(before + 1, after);
 //#endif
     }
+
+    [Fact]
+    public async Task MuxerIsNotCollectedWhenHasBacklog()
+    {
+        // Run on a separate thread to ensure no references to the connection from the main thread
+        var startGC = new TaskCompletionSource<bool>();
+        var testTask = Task.Run(async () =>
+        {
+            using var conn = await ConnectionMultiplexer.ConnectAsync(new ConfigurationOptions()
+            {
+                BacklogPolicy = BacklogPolicy.Default,
+                ConnectTimeout = 50,
+                SyncTimeout = 1000,
+                AllowAdmin = true,
+                EndPoints = { GetConfiguration() },
+            }, Writer);
+            var db = conn.GetDatabase();
+            await db.PingAsync();
+
+            // Disconnect and don't allow re-connection
+            conn.AllowConnect = false;
+            var server = conn.GetServerSnapshot()[0];
+            server.SimulateConnectionFailure(SimulatedFailureType.All);
+            Assert.False(conn.IsConnected);
+
+            var pingTask = Assert.ThrowsAsync<RedisConnectionException>(() => db.PingAsync());
+            startGC.SetResult(true);
+            await pingTask;
+        }).WithTimeout(5000);
+
+        // Use sync wait and sleep to ensure a more timely GC
+        Task.WaitAny(startGC.Task, testTask);
+        while (!testTask.IsCompleted)
```

Comment:
May be worth a timeout exit failure condition here, perhaps after N iterations for some N

Also, to check that it hasn't been collected: really, you need to ask a WeakReference? The test is fine, just: we're not *directly* testing whether it has been collected - we're testing whether the task completed. These are causally related, sure; but if the test name says it is checking collection: we should probably actually check for collection. Or rename the test.

Reply:
Note that testTask actually has a timeout via WithTimeout, and without the GC rooting fix this test times out.

Unless you have some good idea, I'll likely just rename the test because a WeakReference could be used to confirm that the multiplexer is not collected before the ping completes, but could not be used to confirm that it was not collected after that as the GC was free to collect it, and this would require a lot of extra code to avoid strong references.

On the other hand because of the strong reference in using in the body of testTask, the task completing implies that the connection was still alive. I'll add a comment about this at the end of testTask.